### DISCUSSION TITLE
React Native Support

### DISCUSSION THREAD

**Discussion by hattorihanzo86** [pocketbase/pocketbase#2919] [Q&A] [answered]
Hi there is there any support for RN  ?

**Comment by ganigeorgiev** [comment]
The JS SDK should work fine with React Native with some caveats:

- For the realtime subscriptions (and the OAuth2 all-in-one flow) you'll need `EventSource` polyfill, eg.:
  ```js
  // npm install react-native-sse --save
  import eventsource from "react-native-sse";

  global.EventSource = eventsource;
  ```
  _(see https://github.com/pocketbase/js-sdk#nodejs-via-npm)_
  
- The React Native `FormData` implementation on mobile devices is not fully compliant with the web spec version which could cause some issues with file uploads and may require special handling, see https://github.com/pocketbase/pocketbase/discussions/2002#discussioncomment-5254168

- The AuthStore is not persistent and you need to implement your own version depending on how and where do you want to store the data on the device. For one (_untested_) example you can check https://github.com/pocketbase/pocketbase/discussions/1856#discussioncomment-4986746.

**Reply by hattorihanzo86** [reply]
thank you @ganigeorgiev

**Comment by mwaleed-39** [comment]
I faced some issue in React native. I create code for upload image, but it shows error
 WARN  Possible Unhandled Promise Rejection (id: 1):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null

**Reply by ganigeorgiev** [reply]
Without a code sample or at least some details what platform you are targeting and what has triggered the above error, I'm not sure I can help.

You can try to create a minimal reproducible demo with https://snack.expo.dev/ and share it to help with the debugging.

In general, for the file uploads the currently known React Native issue and its workaround is outlined in https://github.com/pocketbase/pocketbase/discussions/2002#discussioncomment-5254168.

**Comment by mwaleed-39** [comment]
CODE:

import React, { useState } from 'react'
import { Alert, Button, View,StyleSheet } from 'react-native'
import * as ImagePicker from 'react-native-image-picker'
import {launchImageLibrary} from 'react-native-image-picker'

const options ={
  title: 'Select Image',
    type: 'library',
    options: {
      maxHeight:200,
      maxWidth:200,
      selectionLimit: 1,
      mediaType: 'photo',
      includeBase64: false,
    }
}

const Request = () => {
   const opengallery = async()=>{
    const image = await ImagePicker.launchImageLibrary(options);
    console.log(image)
   }

  return (
    <View style = {styles.container}>
       <Button title='Upload Image' onPress={opengallery}/>
    </View>
  )
}
export default Request;

const styles = StyleSheet.create({
  container:{
    flex:1,
    alignItems:'center',
    justifyContent:'center'
  }
})

when we pressed upload button, the error message shows:

Possible Unhandled Promise Rejection (id: 1):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null

**Reply by ganigeorgiev** [reply, marked as answer]
I don't think the above error is related to the PocketBase JS SDK (_the PocketBase JS SDK import actually is not even defined in the above_) and it most likely caused by your usage of the `react-native-image-picker` package.

I don't understand why do you need to import it twice, but since the second time you import only the `launchImageLibrary` function, you need to call that function directly as shown in the readme of the library you use - https://github.com/react-native-image-picker/react-native-image-picker#methods (_assuming that this is the one you have in your package.json_). Or in other words, replace ` ImagePicker.launchImageLibrary` with just `launchImageLibrary`.

**Reply by hattorihanzo86** [reply]
Can you make example instance of it? cause sometimes constructor is not called.

**Reply by mwaleed-39** [reply]
CODE:

import React, { useState } from 'react'
import { Alert, Button, View,StyleSheet } from 'react-native'
import * as ImagePicker from 'react-native-image-picker'
import {launchImageLibrary} from 'react-native-image-picker'

const options ={
  title: 'Select Image',
    type: 'library',
    options: {
      maxHeight:200,
      maxWidth:200,
      selectionLimit: 1,
      mediaType: 'photo',
      includeBase64: false,
    }
}

const Request = () => {
   const opengallery = async()=>{
    const image = await ImagePicker.launchImageLibrary(options);
    console.log(image)
   }

  return (
    <View style = {styles.container}>
       <Button title='Upload Image' onPress={opengallery}/>
    </View>
  )
}
export default Request;

const styles = StyleSheet.create({
  container:{
    flex:1,
    alignItems:'center',
    justifyContent:'center'
  }
})

when we pressed upload button, the error message shows:

Possible Unhandled Promise Rejection (id: 1):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null

when we pressed login button the error is Cannot read property 'launchImageLibrary' of null , and unhandled promise rejection

**Reply by ganigeorgiev** [reply]
@mwaleed-39 But you didn't apply the changes that I suggested in https://github.com/pocketbase/pocketbase/discussions/2919#discussioncomment-6461927. Why do you import the `react-native-image-picker` twice in the first place?

In any case, I don't see how your issue is related to PocketBase or the JS SDK. If you have difficulties using the `react-native-image-picker` library, please try to ask for help in the related repo or StackOverflow.

**Reply by mwaleed-39** [reply]
import React, { useState } from 'react'
import { Alert, Button, View, StyleSheet } from 'react-native'

import { launchImageLibrary } from 'expo-image-picker'

const options = {
  title: 'Select Image',
  type: 'library',
  options: {
    maxHeight: 200,
    maxWidth: 200,
    selectionLimit: 1,
    mediaType: 'photo',
    includeBase64: false,
  }
}

const Fine_Repair_Request = () => {
  const opengallery = async () => {
    const image = await launchImageLibrary(options);
    console.log(image)
  }

  return (
    <View style={styles.container}>
      <Button title="Upload Image" onPress={opengallery} />
    </View>
  )
}
export default Fine_Repair_Request;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  }
})

please check it

**Reply by ganigeorgiev** [reply]
@mwaleed-39 I think we are going in circles. It seems that you've replaced `react-native-image-picker` with `expo-image-picker`. If you are going to use the expo package, then please refer to the official documentation and its example in https://docs.expo.dev/versions/latest/sdk/imagepicker/#usage.

But again, your issue is not related to PocketBase or the JS SDK.

Please understand that while I want to help, this is not a general dev support forum and questions like yours will be better addressed in Stack Overflow or in the support channel of the library you use.

**Comment by mwaleed-39** [comment]
Same issue. provide me solution
